### ISSUE TITLE
MA0042 suggestion would result in recursive call

### ISSUE THREAD

**Issue by JeevanJames** [meziantou/Meziantou.Analyzer#246]
**Context**
Rule identifier: _MA0042_
Target Framework: _.NET 5.0_
Language version: _9.0_

**Describe the bug**
I have an base class that has an async and non-async version of a virtual method. Deriving classes can override either method, depending on what they want to do.

```cs
public class Processor
{
    public virtual Task<int> ProcessAsync()
    {
        int result = Process();
        return Task.FromResult(result);
    }
    
    protected virtual int Process()
    {
        throw new NotImplementedException();
    }
}
```

In this case, a MA0042 suggestion is shown for the call to `Process` in the `ProcessAsync` method:
```
MA0042: Use 'ProcessAsync' instead of 'Process'.
```

Using this suggestion will result in a recursive call.

**Comment by meziantou** [comment]
In your case, the fix would introduce a bug. However, I don't think there is a way to _correctly_ avoid false-positives for this analyzer. For instance, it could be perfectly valid to have a recursive algorithm in your case.

So, I don't think there is something to do here except disabling the rule for the false-positive (using `#pragma warning disable MA0042` or `[SuppressMessage]`).
